clients/da-vinci-client/src/main/java/com/linkedin/davinci/notifier/DaVinciPushStatusUpdateTask.java

```diff
@@ -0,0 +1,110 @@
+ */
+public class DaVinciPushStatusUpdateTask {
+  private final Version version;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private boolean batchPushStartSignalSent;
+  private boolean batchPushEndSignalSent;
+  private final long daVinciPushStatusCheckIntervalInMs;
+  private final Map<Integer, ExecutionStatus> partitionStatus = new VeniceConcurrentHashMap<>();
+  // Executor for scheduling tasks
+  private final ScheduledExecutorService scheduler =
+      Executors.newScheduledThreadPool(1, new DaemonThreadFactory("davinci-push-status-update-task-scheduler"));
+
+  public DaVinciPushStatusUpdateTask(
+      Version version,
+      long daVinciPushStatusCheckIntervalInMs,
+      PushStatusStoreWriter pushStatusStoreWriter) {
+    this.version = version;
+    this.batchPushStartSignalSent = false;
+    this.batchPushEndSignalSent = false;
+    this.daVinciPushStatusCheckIntervalInMs = daVinciPushStatusCheckIntervalInMs;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+  }
+
+  public boolean isBatchPushStartSignalSent() {
+    return batchPushStartSignalSent;
+  }
+
+  public boolean isBatchPushEndSignalSent() {
+    return batchPushEndSignalSent;
+  }
+
+  public void batchPushStartSignalSent() {
+    this.batchPushStartSignalSent = true;
+  }
+
+  public void batchPushEndSignalSent() {
+    this.batchPushEndSignalSent = true;
+  }
+
+  public void updatePartitionStatus(int partition, ExecutionStatus status) {
+    partitionStatus.put(partition, status);
+  }
+
+  private void maybeSendBatchingStatus() {
+    // First, check if the batch push start signal has been sent, if not, send out STARTED status for this version
+    // immediately, regardless if there is any partition subscription happened (this is also needed for delayed version
+    // swap for DaVinci during target colo pushes)
+    // If STARTED is sent, and if COMPLETED is not sent, check if all partitions are in COMPLETED status. If so, send
+    // the COMPLETED status.
+    if (!isBatchPushStartSignalSent()) {
+      pushStatusStoreWriter.writeVersionLevelPushStatus(
+          version.getStoreName(),
+          version.getNumber(),
+          ExecutionStatus.STARTED,
+          getTrackedPartitions());
+      batchPushStartSignalSent();
+    } else if (!isBatchPushEndSignalSent() && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED)) {
+      pushStatusStoreWriter.writeVersionLevelPushStatus(
+          version.getStoreName(),
+          version.getNumber(),
+          ExecutionStatus.COMPLETED,
+          getTrackedPartitions());
+      batchPushEndSignalSent();
+      // Shutdown the scheduler after sending the final status
+      shutdown();
+    }
+  }
+
+  /**
+   * Get the partition id set that is being tracked
+   */
+  public Set<Integer> getTrackedPartitions() {
+    return partitionStatus.keySet();
+  }
+
+  public boolean areAllPartitionsOnSameTerminalStatus(ExecutionStatus status) {
```

Comment:
For this, do we want to look at the `VersionBackend#partitionFutures` to make sure every subscribed partition has reported the status before marking job is completed?

Reply:
Yeah, updated in the latest diff.